Jenkinsfile

```diff
@@ -0,0 +1,335 @@
+        }
+      }
+    }
+    
+    /**
+    Build on a linux environment.
+    */
+    stage('build') { 
+      agent { label 'linux && immutable' }
+      environment {
+        PATH = "${env.PATH}:${env.HUDSON_HOME}/go/bin/:${env.WORKSPACE}/bin"
+        GOPATH = "${env.WORKSPACE}"
+      }
+      
+      when { 
+        beforeAgent true
+        environment name: 'linux_ci', value: 'true' 
+      }
+      steps {
+        withEnvWrapper() {
+          unstash 'source'
+          dir("${BASE_DIR}"){    
+            sh """#!/bin/bash
+            ./scripts/jenkins/build.sh
+            """
+          }
+        }
+      }
+    }
+    stage('Parallel stages') {
+      failFast true
+      parallel {
+        stage('test') {
+          agent { label 'linux && immutable' }
+          environment {
+            PATH = "${env.PATH}:${env.HUDSON_HOME}/go/bin/:${env.WORKSPACE}/bin"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          
+          when { 
+            beforeAgent true
+            environment name: 'test_ci', value: 'true' 
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){    
+                sh """#!/bin/bash
+                ./scripts/jenkins/test.sh
+                """
+                codecov('apm-agent-go')
+              }
+            }
+          }
+          post { 
+            always {
+              coverageReport("${BASE_DIR}/build/coverage")
+              junit(allowEmptyResults: true, 
+                keepLongStdio: true, 
+                testResults: "${BASE_DIR}/build/junit-*.xml")
+            }
+          }
+        }
+        stage('Benchmarks') {
+          agent { label 'linux && immutable' }
+          environment {
+            PATH = "${env.PATH}:${env.HUDSON_HOME}/go/bin/:${env.WORKSPACE}/bin"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          
+          when { 
+            beforeAgent true
+            allOf { 
+              //branch 'master';
+              environment name: 'bench_ci', value: 'true' 
+            }
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){    
+                sh """#!/bin/bash
+                ./scripts/jenkins/bench.sh
+                """
+                sendBenchmarks(file: 'build/bench.out', index: "benchmark-go")
+              }
+            }
+          } 
+          post {
+            always {
+              junit(allowEmptyResults: true, 
+                keepLongStdio: true, 
+                testResults: "${BASE_DIR}/build/junit-*.xml")
+            }
+          }
+        }
+        stage('Docker tests') {
+          agent { label 'linux && docker && immutable' }
+          environment {
+            PATH = "${env.PATH}:${env.HUDSON_HOME}/go/bin/:${env.WORKSPACE}/bin"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          
+          when { 
```

Comment:
Can we run docker-test after the parallel stages? I have it like that in Travis CI so that coverage reports aren't published until the unit tests pass on all platforms.

Reply:
in that case, we can move out the test stage, so we will run build, then test, and finally in parallel benchmarks and docker-test

Reply:
SGTM